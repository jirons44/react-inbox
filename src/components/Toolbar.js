import React from 'react';

const Toolbar = ({
         messages,
         onBulkSelected,
         onAddRemoveLabel,
         onDeleteMessages,
         onReadUnreadClicked,
         onToggleComposedForm,
}) => {

    const handleMessagesSelected = () => {
        onBulkSelected( !areAllMessagesSelected() );
    }

    const handleAddLabel = e => {
        if(e.target.value) onAddRemoveLabel(e.target.value, true);
    }

    const handleRemoveLabel = (e) => {
        if(e.target.value) onAddRemoveLabel(e.target.value, false);
    }

    const numOfUnreadMessages = () => {
        return messages.reduce( (sum, message) => {return sum + (message.read ? 0:1)}, 0);
    }

    const areAllMessagesSelected = () => {
        return messages.every(message => message.selected);
    }

    const areSomeMessagesSelected = () => {
        return messages.some(message => message.selected);
    }

    const renderButtonSelectedClassName = () => {
        let className = 'fa fa';

        if (areAllMessagesSelected()) {
            className += '-check'
        } else if (areSomeMessagesSelected()) {
            className += '-minus'
        }

        className += '-square-o';

        return className;
    }

    const selectProps = () => {
        return {
            className: 'form-control label-select',
            disabled:  !areSomeMessagesSelected(),
            value: ''
        }
    }

    const selectOptions = () => {
        return ['dev', 'personal', 'gschool'].map( (item, i)  => <option key={i} value={item}>{item}</option>);
    }

    return (

        <div className="row toolbar">
            <div className="col-md-12">

                <p className="pull-right">
                    <span className="badge badge">{ numOfUnreadMessages() }</span>
                    unread message{numOfUnreadMessages()>1?'s':''}
                </p>

                <a className="btn btn-danger" onClick={onToggleComposedForm}>
                    <i className="fa fa-plus"></i>
                </a>

                <button className="btn btn-default"
                        onClick={ handleMessagesSelected }>
                    <i className={renderButtonSelectedClassName()}></i>
                </button>

                <button className="btn btn-default"
                        disabled={ !areSomeMessagesSelected() }
                        onClick={() => onReadUnreadClicked(true) }>
                    Mark As Read
                </button>

                <button className="btn btn-default"
                        disabled={ !areSomeMessagesSelected() }
                        onClick={()=> onReadUnreadClicked(false) }>
                    Mark As Unread
                </button>

                <select {...selectProps() } onChange={ handleAddLabel }>
                    <option value="">Apply label</option>
                    { selectOptions() }
                </select>

                <select {...selectProps()} onChange={ handleRemoveLabel }>
                    <option value="">Remove label</option>
                    { selectOptions() }
                </select>

                <button className="btn btn-default"
                        disabled={ !areSomeMessagesSelected() }
                        onClick={ onDeleteMessages }>
                    <i className="fa fa-trash-o"></i>
                </button>
            </div>
        </div>
    )
}

export default Toolbar;