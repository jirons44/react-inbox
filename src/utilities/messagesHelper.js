export const getAllIds = messages => {
    if (!messages) return [];
    return messages.map(m => m.id);
 };

export const getByIds = messages => {
    if (!messages) return [];
    return messages.reduce((acc, item) => ({...acc, [item.id]: item,}),{})
}

export const getSelectedMessageIds = ({allIds, byIds}) => {
    if (!allIds && !byIds) return [];
    return allIds.filter(id => byIds[id].selected);
}

export const doesLabelExist = (message, labelName) => {
    let indexOfItem = message.labels.findIndex(label => {return label === labelName;});
    return indexOfItem !== -1;
}

export const toggleSelectedMessage = message => {
    return {...message, selected: !message.selected}
}

