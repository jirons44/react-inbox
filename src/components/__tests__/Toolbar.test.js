import React from 'react';
import renderer from 'react-test-renderer';

import Toolbar  from '../Toolbar';
import MessagesApi from '../../api/mockMessagesApi'
import { getAllIds, getByIds } from '../../utilities/messagesHelper';

const emptyMessages = {
    allIds: [],
    byIds: {}
};

const testMessages = {
    allIds: getAllIds(MessagesApi.getAllMessages()),
    byIds: getByIds(MessagesApi.getAllMessages())
};




test('renders the same way every time', () => {

    //TODO: passing in messages gives invariant messatge
    const tree = renderer.create(
        <Toolbar.WrappedComponent messages={emptyMessages} />)
        .toJSON();

    expect(tree).toMatchSnapshot();
});
