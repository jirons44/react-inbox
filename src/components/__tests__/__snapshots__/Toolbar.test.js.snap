// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renders the same way every time 1`] = `
<div>
  Loading...
</div>
`;
